test/integrationtests/voight_kampff/features/steps/utterance_responses.py

```diff
@@ -0,0 +1,209 @@
+
+
+def dialog_from_sentence(sentence, skill_path):
+    """Find dialog file from example sentence."""
+    dialog_paths = join(skill_path, 'dialog', 'en-us', '*.dialog')
+    best = (None, 0)
+    for path in glob(dialog_paths):
+        patterns = load_dialog_file(path)
+        match, _ = _match_dialog_patterns(patterns, sentence.lower())
+        if match is not False:
+            if len(patterns[match]) > best[1]:
+                best = (path, len(patterns[match]))
+    if best[0] is not None:
+        return basename(best[0])
+    else:
+        return None
+
+
+def _match_dialog_patterns(dialogs, sentence):
+    """Match sentence against a list of dialog patterns.
+
+    Returns index of found match.
+    """
+    # Allow custom fields to be anything
+    d = [re.sub(r'{.*?\}', r'.*', t) for t in dialogs]
+    # Remove left over '}'
+    d = [re.sub(r'\}', r'', t) for t in d]
+    d = [re.sub(r' .* ', r' .*', t) for t in d]
+    # Merge consequtive .*'s into a single .*
+    d = [re.sub(r'\.\*( \.\*)+', r'.*', t) for t in d]
+    # Remove double whitespaces
+    d = ['^' + ' '.join(t.split()) for t in d]
+    debug = 'MATCHING: {}\n'.format(sentence)
+    for i, r in enumerate(d):
+        match = re.match(r, sentence)
+        debug += '---------------\n'
+        debug += '{} {}\n'.format(r, match is not None)
+        if match:
+            return i, debug
+    else:
+        return False, debug
+
+
+@given('an english speaking user')
+def given_english(context):
+    context.lang = 'en-us'
+
+
+@when('the user says "{text}"')
+def when_user_says(context, text):
+    context.bus.emit(Message('recognizer_loop:utterance',
+                             data={'utterances': [text],
+                                   'lang': 'en-us',
+                                   'session': '',
+                                   'ident': time.time()},
+                             context={'client_name': 'mycroft_listener'}))
+
+
+@then('"{skill}" should reply with dialog from "{dialog}"')
+def then_dialog(context, skill, dialog):
+    def check_dialog(message):
+        utt_dialog = message.data.get('meta', {}).get('dialog')
+        return (utt_dialog == dialog.replace('.dialog', ''), '')
+
+    passed, debug = then_wait('speak', check_dialog, context)
+    if not passed:
+        assert_msg = debug
+        assert_msg += mycroft_responses(context)
+
+    assert passed, assert_msg or 'Mycroft didn\'t respond'
+
+
+@then('"{skill}" should reply with "{example}"')
+def then_example(context, skill, example):
+    skill_path = context.msm.find_skill(skill).path
+    dialog = dialog_from_sentence(example, skill_path)
+    print('Matching with the dialog file: {}'.format(dialog))
+    assert dialog is not None, 'No matching dialog...'
+    then_dialog(context, skill, dialog)
+
+
+@then('"{skill}" should reply with anything')
+def then_anything(context, skill):
+    def check_any_messages(message):
+        return (True, '') if message else (False, '')
```

Comment:
what do these tuples represent?  another statement that assigns the tuple to a variable might help with readability.